### PR TITLE
Add inverse Wishart sampler

### PR DESCRIPTION
**Pull request by miaoqingyu2**
Added a sampler from inverse Wishart distribution, as suggested in #5 . I looked into ways to avoid doing a matrix inverse, but it seems like the Bartlett decomposition (which uses Cholesky) is the standard way to go. 

Any comments or suggestions are appreciated!

### CODE REVIEW

**Comment by miaoqingyu2** [comment]
I've updated the `_sample_wishart_bartlett ` function based on your suggestions. Specifically, I renamed the `scale_inv ` argument to scale. I understand the logic behind the change, and in terms of code, that was the only adjustment needed.

Regarding the covariance update in `step_sigma_cov`: I agree it makes sense to use the precision directly and avoid inverting `W`, as you mentioned. My current thought is to handle that in a separate PR, since it will require changes to the `State `class and the initialization step. That said, I'm happy to include it in this PR if you think it's better to keep everything together.

Let me know if anything else needs adjusting, and thanks again for the helpful feedback!

**Comment by Gattocrucco** [comment]
Your changes make sense.

About the inverse: yes most of that stuff with the covariance matrix would go in another PR. However, there is one thing that changes in this PR: that the Wishart sampler should take as input the inverse scale.

Your original code for the Wishart sampler called the input matrix `scale_inv` while it was actually the (non-inverse) scale matrix of the Wishart. Since when using this function you start from the inverse Wishart scale (the inverse Wishart conjugate update on the covariance matrix), you have to modify the function to take as input the inverse scale (for real this time), and use it in the function without ever inverting it.

**Comment by miaoqingyu2** [comment]
I see — after one Gibbs step, the error covariance matrix \( \Sigma \) goes from  
\( \text{Inv-Wishart}(t_0, S_0^{-1}) \)  
to  
\( \text{Inv-Wishart}(t_0 + n, (R R^\top + S_0)^{-1}) \).  

Equivalently, the error precision matrix \( \Sigma^{-1} \) goes from  
\( \text{Wishart}(t_0, S_0) \)  
to  
\( \text{Wishart}(t_0 + n, R R^\top + S_0) \).  

So we want to pass \( R R^\top + S_0 \) to the Wishart sampler and directly sample \( \Sigma^{-1} \), which corresponds to `prec` in your pseudocode example.

**Comment by Gattocrucco** [comment]
I think you got what I mean, though you are still inverting the matrices in your Wishart distributions: you write

$\Sigma\sim\text{Inv-Wishart}(t_0 + n, (R R^\top + S_0)^{-1})$
$\Sigma^{-1}\sim\text{Wishart}(t_0 + n, R R^\top + S_0)$

While it is

$\Sigma\sim\text{Inv-Wishart}(t_0 + n, R R^\top + S_0)$
$\Sigma^{-1}\sim\text{Wishart}(t_0 + n, (R R^\top + S_0)^{-1})$

I wonder—is this because you confuse the two things, or because you consistently use a different notation for the Wishart distribution? I'm following the convention reported on wikipedia (https://en.wikipedia.org/wiki/Inverse-Wishart_distribution, https://en.wikipedia.org/wiki/Wishart_distribution).

**Comment by miaoqingyu2** [comment]
Sorry for the confusion. I was thinking that if I write it this way, we can avoid introducing any explicit matrix inverse, since we just sample directly from $Wishart(t_0+n, R R^T + S_0)$, but I’m fine with either convention.

**Comment by Gattocrucco** [comment]
(Using standard convention) You can't sample from $Wishart(t_0+n, R R^T + S_0)$, that would not make sense as Gibbs step. You have to sample from $Wishart(t_0+n, (R R^T + S_0)^{-1})$. That does not imply that in the actual implementation you have to invert the matrix, you can write a function that samples the wishart starting from the inverse scale. It's like... if I write $z = x+y$, I could implement it in Python (for whatever reasons) as

```py
neg_y = ...
def weird_sum(x, neg_y):
    return x - neg_y # actually a subtraction here!
z = weird_sum(x, neg_y) # z = x + y
```

File: src/bartz/mcmcstep.py [hunk 1]
```diff
@@ -2473,6 +2474,43 @@ def step_sigma(key: Key[Array, ''], bart: State) -> State:
...
+        Degrees of freedom
+    scale_inv
+        Scale matrix of the corresponding Inverse Wishart distribution
+
+    Returns
+    -------
+    A sample from Wishart(df, scale)
+    """
+    k = scale_inv.shape[0]
+    L = jnp.linalg.cholesky(scale_inv)
```
Comment (Gattocrucco) [inline src/bartz/mcmcstep.py:2497]: Cholesky fails if the matrix is not positive definite. Numerically, this can happen even if the matrix if positive definite on paper, due to numerical approximation error.

I don't know how much it is likely that this would happen here; there would need to be strongly correlated directions in the residuals. In any case, I want it preemptively handled since it's a common problem.

To handle it, use Gershgorin's circle theorem to obtain an estimate of the largest eigenvalue, then use it to estimate the numerical error, and add a multiple of the identity to the matrix to decompose such that all eigenvalues are at least as large as the numerical error:

```py
rho = max(sum(abs(A), axis=1))  # upper bound on eigenvalues with Gershgorin's circle theorem
u = len(A) * rho * jnp.finfo(A.dtype).eps  # estimate of the numerical error
A = A.at[jnp.diag_indices_from(A)].add(u)  # inflate all eigenvalues by the numerical error
L = cholesky(A)  # now the Cholesky decomposition should not fail even if is A is close to singular
```

File: src/bartz/mcmcstep.py [hunk 2]
```diff
@@ -2473,6 +2474,43 @@ def step_sigma(key: Key[Array, ''], bart: State) -> State:
...
+
+    # Diagonal elements: A_ii ~ sqrt(chi^2(df - i))
+    # chi^2(k) = Gamma(k/2, scale=2)
+    df_vector = df - jnp.arange(k)
+    chi2_samples = random.gamma(diag_key, df_vector / 2.0) * 2.0
+    diag_A = jnp.sqrt(chi2_samples)
+
+    off_diag_A = random.normal(offdiag_key, (k, k))
+    A = jnp.tril(off_diag_A, -1) + jnp.diag(diag_A)
+    T = solve_triangular(L, A, lower=True, trans='T')
```
Comment (Gattocrucco) [inline src/bartz/mcmcstep.py:2509]: I did not know `solve_triagular` has a `trans` option! I shall use that in the future.

File: src/bartz/mcmcstep.py [hunk 3]
```diff
@@ -2473,6 +2474,43 @@ def step_sigma(key: Key[Array, ''], bart: State) -> State:
...
+
+    return T @ T.T
```
Comment (Gattocrucco) [inline src/bartz/mcmcstep.py:2511]: I checked the calculation you use and I think it's correct.

File: tests/test_mvbart.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+from jax import numpy as jnp
+from jax import random
+from numpy.testing import assert_allclose
+
+from bartz.mcmcstep import _sample_wishart_bartlett
+
+
+def test_wishart_bartlett_output():
+    """Test the basic properties of the wishart sampler output."""
+    key = random.key(123)
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:12]: I have pre-defined a pytest fixture for random keys.

To use it, add a `keys` argument to `test_wishart_bartlett_output`, and then do `keys.pop()` every time you need a key.

You can for example look at `test_mcmcstep.py` to see examples of this.

The way this works is that pytest recognizes the `keys` argument, and notices it has the same name as the `keys` fixture in `tests/conftest.py`. So it calls the fixture function to generate `keys` and passes it as argument to your function.

The reason I use this system is such that the keys are different and unique for each test case (including parametrizations), but they are also deterministic (always the same key for the same test case every time you run the test suite). Compared to, say, `random.key(0)` or `random.key(123)`, this practice makes it more likely to catch bugs that appear only for some rare input values. It's also more convenient because you can just add `keys` to the function signature and use it.

File: tests/test_mvbart.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+"""Test multivariate BART components."""
+
+from jax import numpy as jnp
+from jax import random
+from numpy.testing import assert_allclose
+
+from bartz.mcmcstep import _sample_wishart_bartlett
+
+
+def test_wishart_bartlett_output():
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:10]: Reorganize this test case by splitting in three separate test cases in a class, like:

```py
class TestWishart:
    def random_pd_matrix(key, k): ...  # re-use this method in all test cases
    def test_size(self, keys): ...
    def test_symmetric(self, keys): ...
    def test_pos_def(self, keys): ...
```

File: tests/test_mvbart.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    k = 3
+    # Create a random positive definite matrix for the scale
+    A = random.normal(scale_key, (k, k))
+    scale = A @ A.T + jnp.eye(k)
+
+    wishart_sample = _sample_wishart_bartlett(sample_key, df, scale)
+
+    # Check shape
+    assert wishart_sample.shape == (k, k)
+    assert_allclose(wishart_sample, wishart_sample.T, rtol=1e-6)
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:24]: Instead of `np.testing.assert_allclose`, use `tests.util.assert_close_matrices`.

It's a function I wrote for comparing vectors and matrices. It's useful because with many elements sometimes there are outliers, and putting a threshold that's fine with outliers makes the test less sensitive. This in practice pops up when I run the unit tests on a different system/machine with slightly different floating point results, or the random keys change, and some thresholds that were previously fine become too tight.

`assert_close_matrices` uses the whole matrix norm to do the comparison, instead of comparing each element one-by-one.

File: tests/test_mvbart.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    # Check for positive definiteness (eigenvalues should be positive)
+    eigenvalues = jnp.linalg.eigvalsh(wishart_sample)
+    assert jnp.all(eigenvalues > 0)
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:28]: All your test cases check for basic properties of a Wishart rv, but there is no test that checks the distribution of the samples is correct. Add one.

I don't know what's the most convenient way to check the distribution is correct, I have only written tests for univariate distributions so far (like `test_jaxext.py::TestTruncatedNormalOneSided::test_correct`), those are easier.

File: tests/test_mvbart.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+def test_wishart_bartlett_output():
+    """Test the basic properties of the wishart sampler output."""
+    key = random.key(123)
+    scale_key, sample_key = random.split(key)
+    df = 5
+    k = 3
+    # Create a random positive definite matrix for the scale
+    A = random.normal(scale_key, (k, k))
+    scale = A @ A.T + jnp.eye(k)
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:18]: This way of generating p.d. matrices produces matrices which are well-conditioned, due to adding the identity. You should also produce matrices that are singular or close to singular (i.e., just positive semi-definite instead of p.d.), to stress test the "matrix inversion" (`solve_triangular`) in the implementation.

File: tests/test_mvbart.py (hunk 7 above)
Comment (miaoqingyu2) [reply tests/test_mvbart.py]: 've added a distribution test based on the property that: if $M \sims W_p(n,\Sigma)$ and $a\in\mathbb{R}^p$ is such that $a^T M a \neq 0$, then 
$$
\frac{a^T M a}{a^T \Sigma a} \sims \chi^2_n
$$

Comment (miaoqingyu2) [reply tests/test_mvbart.py]: For the distribution test I used inv(). My understanding is that it’s okay here since the inverse isn’t propagated in the Gibbs sampler step, but I’m happy to change the implementation if you’d prefer a formulation that avoids the explicit inverse.

File: tests/test_mvbart.py [hunk 9]
```diff
@@ -2,27 +2,80 @@
...
+            else:
+                smalls = jnp.geomspace(1.0, 1e-6, num=k - 1)
+                eigs = jnp.concatenate([smalls, jnp.array([0.0])])
+        else:
+            eigs = jnp.geomspace(1.0, 1.0 / condition_number, num=k)
+        return (U @ jnp.diag(eigs)) @ U.T
+
+    def test_size(self, keys):
+        """Check that the sample generated by wishart sampler is of shape k*k."""
+        df, k = 5, 3
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:36]: The sampler should be tested for multiple values of k. I'd do `[1, low value, large value]`, say `[1, 3, 100]`.

To automatize this kind of thing, pytest provides parametrization, and even parametrized fixtures. This means you can do it like this:

```python
class TestFoo:
    
    @pytest.fixture(params=['a', 'b', 'c'])
    def something(request) -> str:
        return request.param
    
    def test_bar(self, something: str):
        if something == 'a':
             ...
        elif something == 'b':
             ...
        else:
             ...

    def test_baz(self, something: str):
        ...
```

This makes `test_bar` and `test_baz` automatically repeat 3 times with the three possible values of `something`.

File: tests/test_mvbart.py (hunk 7 above)
Comment (Gattocrucco) [reply tests/test_mvbart.py]: I agree the inverse here is probably fine. You are inverting a matrix that you know to be well conditioned, and it's a unit test so you have control over everything.

It would make sense to use something more sophisticated only if a unit test wasn't working and you suspected the lowered numerical accuracy due to the inverse could be a problem.

Comment (Gattocrucco) [reply tests/test_mvbart.py]: Good idea to use that property of the Wishart.

**Review by Gattocrucco — CHANGES_REQUESTED** [review]
Good, thanks. I marked a few more things to do.

To be honest, I expect the Wishart sampler to be already fine as of now. I'm going overboard with doing everything well as an excuse to make sure you know a few more things about how to write good tests, before you go on to do the more complex stuff.

File: tests/test_mvbart.py [hunk 10]
```diff
@@ -2,27 +2,80 @@
...
+        """Check that the wishart sampler still works with singular or near singular matrix."""
+        df, k = 5, 3
+        ill_conditioned_scale = self.ill_conditioned_matrix(keys.pop(), k, 1e6, True)
+        sample = _sample_wishart_bartlett(keys.pop(), df, ill_conditioned_scale)
+        assert jnp.all(jnp.isfinite(sample))
+
+    def test_wishart_dist(self, keys):
+        """Check that the sample generated by wishart sampler follows a wishart distribution."""
+        df, k = 5, 3
+        nsamples = 100
```
[the remark this answers is not in the document]
Comment (miaoqingyu2) [reply tests/test_mvbart.py:66]: i am picking nsamples = 100 here because keys fixture only creates 128 keys, while i need to use a key each time in the for loop. Now  i have changed the keys in confest accordingly, hope that's ok.

[the remark this answers is not in the document]
Comment (miaoqingyu2) [reply tests/test_mvbart.py:66]: it takes ~30s on my end when i change nsamples = 1000

File: tests/conftest.py [hunk 11]
```diff
@@ -49,4 +49,4 @@ def keys(request):
     rng = np.random.default_rng(seed)
     seed = np.array(rng.bytes(4)).view(np.uint32)
     key = jax.random.key(seed)
-    return jaxext.split(key, 128)
+    return jaxext.split(key, 1024)
```
Comment (Gattocrucco) [inline tests/conftest.py:52]: Set this back to 128 (see later comment for why).

File: tests/test_mvbart.py [hunk 12]
```diff
@@ -11,69 +12,76 @@
...
         sigma = self.random_pd_matrix(keys.pop(), k)
         scale_inv = jnp.linalg.inv(sigma)
 
         a = random.normal(keys.pop(), (k,))
-        denumerator = float(a.T @ sigma @ a)
+        denumerator = a.T @ sigma @ a
 
         ts = []
         for _ in range(nsamples):
             W = _sample_wishart_bartlett(keys.pop(), df, scale_inv)
```
Comment (Gattocrucco) [inline tests/test_mvbart.py:83]: Do

```py
for key in random.split(keys.pop(), nsamples):
    W = _sample_wishart_bartlett(key, df, scale_inv)
```

The reason is that in general `nsamples` could be changed arbitrarily, so it's convenient to decouple this from the pre-set number of keys in `conftest.py`.

So when there's a large array of keys to generate, use `random.split(keys.pop(), ...)` instead of using `keys.pop()` many times.

File: src/bartz/mcmcstep.py [hunk 13]
```diff
@@ -2499,7 +2499,7 @@ def _sample_wishart_bartlett(
 
     # Gershgorin estimate for max eigenvalue
     rho = jnp.max(jnp.sum(jnp.abs(scale_inv), axis=1))
-    u = k * rho * jnp.finfo(scale_inv.dtype).eps
+    u = k * rho * jnp.finfo(scale_inv.dtype).eps + jnp.finfo(scale_inv.dtype).eps
```
Comment (Gattocrucco) [inline src/bartz/mcmcstep.py:2502]: Yeah adding an "absolute" epsilon to the "relative" one seems a good idea.

Did you add this because of some problems with the tests? Or just on first principles?

File: tests/test_mvbart.py (hunk 12 above)
Comment (Gattocrucco) [reply tests/test_mvbart.py:83]: Since you mentioned that it's slow, I'd do the following things:

1. Decorate `_sample_wishare_bartlett` with `jit`

2. If it's still slow, make a version of `_sample_wishart_bartlett` just for this test case with `jax.vmap`:

```py
...
_sample_vectorized = vmap(_sample_wishart_bartlett, in_axes=(0, None, None))
Ws = sample_vectorized(random.split(keys.pop(), nsamples), ...)
# now Ws should be a (nsamples, k, k) array with all the samples
```

**Review by Gattocrucco — CHANGES_REQUESTED** [review]

File: src/bartz/mcmcstep.py (hunk 13 above)
Comment (miaoqingyu2) [reply src/bartz/mcmcstep.py:2502]: i added the second term because when testing k = 1 case, rho = 0 therefore the first term of u = 0

**Review by Gattocrucco — APPROVED** [review]
All green! Please confirm to me there's nothing else you would like to change yourself, then I'll merge.
